### PR TITLE
Build Ubuntu 22.04 with ROS humble

### LINKED ISSUE #40: Ubuntu 22.04 version of the RPI4 image
Hi,

I am assuming ubuntu 20.04 is currently used for the rpi4 image
In this post the ubuntu devs explain how interacting with the GPIO pins has changed since 21.04:
https://ubuntu.com/tutorials/gpio-on-raspberry-pi#1-overview

I'd like to use 22.04 going forward so I don't have to rewrite my code which uses GPIO pins. Is 22.04 already currently being looked at?

### PR DESCRIPTION
**Pull request by shuhaowu**
This is basically a port of focal-rt-galactic to jammy-rt-focal by copy pasting the files. It uses the 5.15 kernel from https://github.com/ros-realtime/linux-real-time-kernel-builder.

Resolves #40

### CODE REVIEW

File: image_builder/data/jammy-rt-humble/scripts/phase1-target [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Setup ROS distro and ROS
+sudo curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key  -o /usr/share/keyrings/ros-archive-keyring.gpg
```
Comment (LanderU) [inline image_builder/data/jammy-rt-humble/scripts/phase1-target:4]:
```suggestion
curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key -o /usr/share/keyrings/ros-archive-keyring.gpg
```

Comment (LanderU) [reply image_builder/data/jammy-rt-humble/scripts/phase1-target:4]: Do we need `sudo`?

File: image_builder/data/jammy-rt-humble/scripts/phase1-target [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+sudo apt-get update
```
Comment (LanderU) [inline image_builder/data/jammy-rt-humble/scripts/phase1-target:5]:
```suggestion
apt-get update
```

File: image_builder/data/jammy-rt-humble/scripts/phase1-target [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+sudo apt-get install -y ros-$ROS_DISTRO-ros-base
```
Comment (LanderU) [inline image_builder/data/jammy-rt-humble/scripts/phase1-target:6]:
```suggestion
apt-get install -y ros-$ROS_DISTRO-ros-base
```

File: image_builder/data/focal-rt-galactic/scripts/phase1-target [hunk 4]
```diff
@@ -1,9 +1,9 @@
 #!/bin/bash
 
 # Setup ROS distro and ROS
-sudo curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key  -o /usr/share/keyrings/ros-archive-keyring.gpg
-sudo apt-get update
-sudo apt-get install -y ros-$ROS_DISTRO-ros-base
+curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key  -o /usr/share/keyrings/ros-archive-keyring.gpg
```
Comment (LanderU) [inline image_builder/data/focal-rt-galactic/scripts/phase1-target:4]:
```suggestion
curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key -o /usr/share/keyrings/ros-archive-keyring.gpg
```

**Comment by shuhaowu** [comment]
Following the upstream kernel release, I have finished this port. I haven't tested the image tho, but will do at some point in the weekend.

**Comment by shuhaowu** [comment]
Image boots into the rt kernel. Going to let cyclictest run overnight and will report back in some hours...

```
ubuntu@ubuntu:~$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04 LTS
Release:        22.04
Codename:       jammy
ubuntu@ubuntu:~$ uname -a
Linux ubuntu 5.15.39-rt42-raspi #1 SMP PREEMPT_RT Sat Jul 16 08:34:58 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
ubuntu@ubuntu:~$ echo $ROS_DISTRO 
humble
ubuntu@ubuntu:~$ ros2
usage: ros2 [-h] [--use-python-default-buffering] Call `ros2 <command> -h` for more detailed usage. ...

ros2 is an extensible command-line tool for ROS 2.

...
```

File: image_builder/data/jammy-rt/scripts/phase1-target [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  mdadm \
+  snapd \
+  btrfs-progs \
+  xfsprogs
+
+# TODO: Even tho
+# https://github.com/ros-realtime/linux-real-time-kernel-builder/pull/32 is
+# merged, this is still needed somehow... I suspect that flash-kernel is no
+# longer being called due to some changes upstream? Not sure.
+ln -s -f /boot/vmlinuz-${LINUX_RT_VERSION}-raspi /boot/vmlinuz
```
Comment (LanderU) [inline image_builder/data/jammy-rt/scripts/phase1-target:38]: hmm, we need to check ths

**Review by LanderU — APPROVED** [review]
LGTM!

**Comment by LanderU** [comment]
Thanks a lot @shuhaowu!

**Comment by carlossvg** [comment]
@shuhaowu I tested it too and it seems to be working fine. I created a pre-release here: https://github.com/ros-realtime/ros-realtime-rpi4-image/releases/tag/ubuntu-22.04-rt-ros2-humble-arm64%2Braspi

I decided to add a cyclictest plot as part of the release info. If you want you can replace the plot image with the one you got from your tests.

**Review by carlossvg — APPROVED** [review]

**Comment by shuhaowu** [comment]
Good idea to have a plot for each release. I have the numbers and data, but I need to plot it. Here's a preview of the data (2 hours of cyclictest + stress-ng -c 4), it seems that 5.15 is significantly better that 5.4 (~70us vs ~130us):

```
ubuntu@ubuntu:~$ sudo cyclictest --mlockall --smp --priority=80 --interval=200 --distance=0 -D 120m -H 400 --histfile=cyclictest.log
WARN: distance is ignored and set to 0, if histogram enabled
WARN: stat /dev/cpu_dma_latency failed: No such file or directory
policy: fifo: loadavg: 6.67 6.56 6.49 6/169 1049           

T: 0 (  950) P:80 I:200 C:35999959 Min:     13 Act:   15 Avg:   14 Max:      47
T: 1 (  951) P:80 I:200 C:35999827 Min:     13 Act:   14 Avg:   15 Max:      47
T: 2 (  952) P:80 I:200 C:35999813 Min:     13 Act:   14 Avg:   15 Max:      56
T: 3 (  953) P:80 I:200 C:35999752 Min:     13 Act:   18 Avg:   16 Max:      69
```

For the plots, I'm thinking that I should create a set of scripts that tests and plots in a standard manner. I'll do that, PR it into this repo, and then update the plots in the release.

File: image_builder/data/jammy-rt/scripts/phase1-target (hunk 5 above)
Comment (shuhaowu) [reply image_builder/data/jammy-rt/scripts/phase1-target:38]: Yeah. Need to create an issue upstream again, maybe.

**Comment by carlossvg** [comment]
@shuhaowu Maybe you can use this one https://github.com/ros-realtime/ros_realtime_benchmarks_config/tree/main/cyclictest

**Comment by shuhaowu** [comment]
I have this one: https://github.com/cactusdynamics/rt-demo/blob/master/data/cyclictest-rpi4/plot.ipynb. This plotter allows me to do comparisons. I'm porting it into a small library and have it on a jupyter notebook here.

**Pull request merged**